drivers/char/tpm/tpm_ftpm_optee.c

```diff
@@ -0,0 +1,304 @@
+	int rc;
+
+	mutex_lock(&tpm_tee_context.response_buffer_lock);
+
+	response_length = tpm_tee_context.response_length;
+	if (len < response_length) 
+	{
+		rc = -EIO;
+		goto cleanup;
+	}
+
+	memcpy(buf, tpm_tee_context.response_buffer, response_length);
+	tpm_tee_context.response_length = 0;
+
+	rc = response_length;
+
+cleanup:
+	mutex_unlock(&tpm_tee_context.response_buffer_lock);
+	return rc;
+}
+
+static int ftpm_optee_tpm_op_send(struct tpm_chip *chip, u8 *buf, size_t len)
+{
+	int rc = 0;
+	size_t response_length;
+	u8* response_buffer;
+	struct tpm_output_header* response_header;
+	struct tee_param submit_command_params[TEEC_CONFIG_PAYLOAD_REF_COUNT] = {
+		{ .attr = TEE_IOCTL_PARAM_ATTR_TYPE_MEMREF_INPUT },
+		{ .attr = TEE_IOCTL_PARAM_ATTR_TYPE_MEMREF_INOUT },
+		{ .attr = TEE_IOCTL_PARAM_ATTR_TYPE_NONE },
+		{ .attr = TEE_IOCTL_PARAM_ATTR_TYPE_NONE }
+	};
+	struct tee_ioctl_invoke_arg submit_command_args;
+	struct tee_shm *shm = NULL;
+
+	mutex_lock(&tpm_tee_context.response_buffer_lock);
+
+	tpm_tee_context.response_length = 0;
+
+	submit_command_args.func = FTPM_OPTEE_TA_SUBMIT_COMMAND;
+	submit_command_args.session = tpm_tee_context.optee_session_id;
+	submit_command_args.num_params = TEEC_CONFIG_PAYLOAD_REF_COUNT;
+
+	shm = tee_shm_alloc(
+		tpm_tee_context.ftpm_optee_context, 
+		len + sizeof(tpm_tee_context.response_buffer), 
+		TEE_SHM_MAPPED | TEE_SHM_DMA_BUF);
+
+	if (IS_ERR(shm))
+	{
+		rc = -ENOMEM;
+		goto cleanup;
+	}
+
+	// request
+	submit_command_params[0].u.memref.size = len;
+	submit_command_params[0].u.memref.shm = shm;
+	submit_command_params[0].u.memref.shm_offs = 0;
+	memcpy(submit_command_params[0].u.memref.shm->kaddr, buf, len);
+
+	// response
+	submit_command_params[1].u.memref.size = sizeof(tpm_tee_context.response_buffer);
+	submit_command_params[1].u.memref.shm = shm;
+	submit_command_params[1].u.memref.shm_offs = len;
+
+	rc = tee_client_invoke_func(
+		tpm_tee_context.ftpm_optee_context, 
+		&submit_command_args, 
+		submit_command_params);
+	if (rc || submit_command_args.ret)
+	{
+		rc = rc ? rc : submit_command_args.ret;
+		goto cleanup;
+	}
+
+	response_buffer = submit_command_params[1].u.memref.shm->kaddr + 
+		submit_command_params[1].u.memref.shm_offs;
+	response_header = (struct tpm_output_header*)response_buffer;
+	response_length = be32_to_cpu(response_header->length);
+
+	if (response_length > sizeof(tpm_tee_context.response_buffer))
+	{
+		rc = -EIO;
+		goto cleanup;
+	}
+
+	memcpy(
+		tpm_tee_context.response_buffer, 
+		response_buffer, 
+		response_length);
+	tpm_tee_context.response_length = response_length;
+
+cleanup:
+	if (shm)
+	{
+		tee_shm_free(shm);
+	}
+
+	mutex_unlock(&tpm_tee_context.response_buffer_lock);
+
+	return rc;
+}
+
+static void ftpm_optee_tpm_op_cancel(struct tpm_chip *chip)
+{
+	;
+}
+
+static u8 ftpm_optee_tpm_op_status(struct tpm_chip *chip)
+{
+	return 0;
+}
+
+static bool ftpm_optee_tpm_req_canceled(struct tpm_chip *chip, u8 status)
+{
+	return 0;
```

Comment:
Same as previous.

Reply:
Just FYI, you can use TEE Cancellation API (this will need to add additional wrapper for `ctx->teedev->desc->ops->cancel_req` in "tee: add kernel internal client interface ") to cancel the previously sent command.

and within MS TPM TA in `platform/Cancel.c` you can check cancellation flag by `TEE_GetCancellationFlag()`.

Reply:
Thanks for the pointer Igor, will look at it for the next round of changes.